File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-9.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # [3,4,5,6,1,2]
        # [1, 2, 3, 4, 5, 6]
        # [6, 1, 2, 3, 4, 5]
        # [3, 4, 5, 6][1, 2]
        #  l              r
        # sum < target l++
        # r--
        # if nums[mid] <= currMin: currMin = mid
        #[1]
        # nums[mid] <= nums[r]: r = mid - 1
        # else: l = mid + 1

        # other approach:
        l = 0
        r = len(nums) - 1

        while l < r:

            mid = l + (r - l) // 2

            # Case 1: Minimum is in right half
            if nums[mid] > nums[r]:
                l = mid + 1

            # Case 2: Minimum is at mid or in left half
            else:
                r = mid

        return nums[l]


        # while l <= r:
        #     if nums[l] < nums[r]: # feasible?
        #         res = min(res, nums[l])
            
        #     mid = l + (r - l) // 2
        #     res = min(res, nums[mid])
        #     if nums[mid] > nums[r]:
        #         l = mid + 1
        #     else:
        #         r = mid - 1
        # return res

            

File: Data Structures & Algorithms/three-integer-sum/submission-9.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // nums = [-1,0,1,2,-1,-4] -> nlogn
        //[-4,-1,-1, 0,1,2]
        //for i -> nums
        //if(nums[i] > 0) continue
        //if(i > 0 && nums[i] == nums[i-1]) continue;
        //l = i + 1; r = nums.length - 1
        //while(l < r) ->
        // sum = nums[i] + nums[l] + nums[r]
        //if sum < target l++
        //else if sum > target r--
        //else -> 
        //list.add(Arrays.asList(nums[i] , nums[l] , nums[r]))
        //l++, r--
        //while(l < r && nums[l] == nums[l -1])l++

        List<List<Integer>> res = new ArrayList<>();//O(m)
        Arrays.sort(nums);//O(nlogn) + O(n^2)
        int target = 0;
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] > target) break;
            if(i > 0 && nums[i] == nums[i-1]) continue;

            int l = i + 1; 
            int r = nums.length - 1;

            while(l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if(sum < target) l++;
                else if(sum > target) r--;
                else {
                    res.add(Arrays.asList(nums[i],nums[l],nums[r]));
                    l++;
                    r--;
                    while(l < r && nums[l] == nums[l-1])l++;
                }
            } 
        }  

        return res;
    }
}
